File: agent-app-go/internal/answer/structured_evidence_formatter.go
// Package answer ports com.riwonace.agent.answer from the Kotlin/Spring AI baseline.
package answer

import (
	"regexp"
	"strings"
)

var (
	// singleCountRe is anchored to mirror Kotlin's matchEntire (the whole trimmed
	// string must match, not just a substring).
	singleCountRe      = regexp.MustCompile(`(?i)^(?:count|count\(\*\))=([0-9]+)$`)
	explicitCountUnitRe = regexp.MustCompile(`몇\s*(명|개|건|곳)`)
)

// StructuredEvidenceFormatter ports StructuredEvidenceFormatter.kt 1:1.
type StructuredEvidenceFormatter struct{}

func (f *StructuredEvidenceFormatter) FormatSQL(question, evidence string) string {
	m := singleCountRe.FindStringSubmatch(strings.TrimSpace(evidence))
	if m == nil {
		return evidence
	}
	count := m[1]
	return "총 " + count + countUnit(question) + " 입니다."
}

func countUnit(question string) string {
	if m := explicitCountUnitRe.FindStringSubmatch(question); m != nil {
		return m[1]
	}
	switch {
	case strings.Contains(question, "건"):
		return "건"
	case strings.Contains(question, "직원"), strings.Contains(question, "사람"), strings.Contains(question, "인원"):
		return "명"
	case strings.Contains(question, "고객사"):
		return "곳"
	default:
		return "개"
	}
}

File: agent-app-ktor/src/main/kotlin/com/riwonace/agentktor/answer/StructuredEvidenceFormatter.kt
package com.riwonace.agentktor.answer


/** 구조화 MCP 결과의 값을 바꾸지 않고 질문에 드러난 표현 단위만 복원한다. */
class StructuredEvidenceFormatter {
    fun formatSql(question: String, evidence: String): String {
        val count = SINGLE_COUNT.matchEntire(evidence.trim())?.groupValues?.get(1) ?: return evidence
        return "총 $count${countUnit(question)} 입니다."
    }

    private fun countUnit(question: String): String {
        EXPLICIT_COUNT_UNIT.find(question)?.groupValues?.get(1)?.let { return it }
        return when {
            question.contains("건") -> "건"
            question.contains("직원") || question.contains("사람") || question.contains("인원") -> "명"
            question.contains("고객사") -> "곳"
            else -> "개"
        }
    }

    companion object {
        private val SINGLE_COUNT = Regex("(?i)(?:count|count\\(\\*\\))=([0-9]+)")
        private val EXPLICIT_COUNT_UNIT = Regex("몇\\s*(명|개|건|곳)")
    }
}
